### PR TITLE
Refactor out state delta handling into its own class

### PR DESCRIPTION
**Pull request by hawkowl**
The new stats module uses some of the common components, so it's broken up into its own module.

### CODE REVIEW

File: synapse/handlers/user_directory.py [hunk 1]
```diff
@@ -26,10 +26,12 @@
 from synapse.types import get_localpart_from_id
 from synapse.util.metrics import Measure
 
+from .state_deltas import StateDeltasHandler
```
Comment (richvdh) [inline synapse/handlers/user_directory.py]: generally we prefer absolute imports

File: synapse/handlers/user_directory.py [hunk 2]
```diff
@@ -26,10 +26,12 @@
...
+
 logger = logging.getLogger(__name__)
 
 
-class UserDirectoryHandler(object):
+class UserDirectoryHandler(StateDeltasHandler):
```
Comment (richvdh) [inline synapse/handlers/user_directory.py]: composition > inheritance for this sort of thing, imho

File: synapse/storage/state_deltas.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+from ._base import SQLBaseStore
+
+logger = logging.getLogger(__name__)
+
+
+class StateDeltasStore(SQLBaseStore):
+
+    @defer.inlineCallbacks
+    def get_all_rooms(self):
```
Comment (richvdh) [inline synapse/storage/state_deltas.py]: this is new, and unused so far? Not sure it is related to StateDeltas anyway.

File: synapse/storage/state_deltas.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        """Get all room_ids we've ever known about, in ascending order of "size"
+        """
+        sql = """
+            SELECT room_id FROM current_state_events
```
Comment (richvdh) [inline synapse/storage/state_deltas.py]: not sure if this is the best way to get a list of rooms

File: synapse/storage/state_deltas.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            GROUP BY room_id
+            ORDER BY count(*) ASC
+        """
+        rows = yield self._execute("get_all_rooms", None, sql)
+        defer.returnValue([room_id for room_id, in rows])
+
+    @defer.inlineCallbacks
+    def get_all_local_users(self):
```
Comment (richvdh) [inline synapse/storage/state_deltas.py]: this seems unrelated to StateDeltas. IOW: what are the criteria for things being in this class rather than another? If there is a good reason, please make it explicit with a docstring on the class.

File: synapse/handlers/state_deltas.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import logging
+
+from twisted.internet import defer
+
+from ._base import BaseHandler
+
+logger = logging.getLogger(__name__)
+
+
+class StateDeltasHandler(BaseHandler):
```
Comment (richvdh) [inline synapse/handlers/state_deltas.py]: I keep hoping that `BaseHandler` will go away and die. Unless there's a need to inherit from it I'd prefer it if we didn't?

File: synapse/storage/state_deltas.py (hunk 4 above)
Comment (hawkowl) [reply synapse/storage/state_deltas.py]: Sure, but it's moved existing code.

File: synapse/handlers/user_directory.py (hunk 2 above)
Comment (hawkowl) [reply synapse/handlers/user_directory.py]: sure, that could be a good refactor later. Right now, I'm just splitting up what Matthew wrote.

File: synapse/storage/state_deltas.py (hunk 3 above)
Comment (hawkowl) [reply synapse/storage/state_deltas.py]: yeah, it was a common bit (no longer used) that was split out.

File: synapse/storage/state_deltas.py (hunk 4 above)
Comment (hawkowl) [reply synapse/storage/state_deltas.py]: (which actually no longer exists in user_directory on develop)

File: synapse/storage/state_deltas.py (hunk 3 above)
Comment (richvdh) [reply synapse/storage/state_deltas.py]: if it's no longer used, kill it rather than re-adding it?

File: synapse/handlers/user_directory.py (hunk 2 above)
Comment (richvdh) [reply synapse/handlers/user_directory.py]: I don't think it will happen later. I'm not too bothered about this particular nit, but as a general principle, I don't think that we should be merging stuff into develop which wouldn't normally pass review just because Matthew wrote it in a hurry. 

I'm sorry if you've been given a pile of code to try and sort out, but I am strongly averse to adding to our technical debt by adding essentially unreviewed code to develop.

File: synapse/handlers/state_deltas.py (hunk 6 above)
Comment (hawkowl) [reply synapse/handlers/state_deltas.py]: don

File: synapse/storage/state_deltas.py (hunk 5 above)
Comment (hawkowl) [reply synapse/storage/state_deltas.py]: yeah, it is. and: I don't know, really, but the state delta handling code seems to be what's left.

File: synapse/storage/state_deltas.py (hunk 3 above)
Comment (hawkowl) [reply synapse/storage/state_deltas.py]: it's dead now

File: synapse/handlers/user_directory.py (hunk 2 above)
Comment (hawkowl) [reply synapse/handlers/user_directory.py]: Sure, that's why this code is being reviewed. And, yes, I am a staunch believer in the composition > inheritance thing, which -- if I wrote it -- I might well have investigated how to do it. But, I didn't write it, but the code works, and it matches the rest of the inheritance-based storage/handler APIs, so there's no reason going back and thinking about how to make this API composition-friendly if we're not doing it across everything else that's inheritance based.

Comment (richvdh) [reply synapse/handlers/user_directory.py]: my impression was that changing to composition here would be approximately a three line change, but if it's more complicated than that, I'm happy to concede.
